File: rod-js/src/builders.ts
import { Bridge } from "./bridge-types";

export type ParseOptions = {
    /**
     * 'lazy': Uses JS reflection (Reflect.get) to traverse data. 
     * Best for large objects where only a few fields are validated.
     * 
     * 'eager': Serializes entire input to Rust JSON structure first. 
     * Best for heavy validation logic on smaller objects or massive arrays of primitives.
     * 
     * @default "lazy"
     */
    mode?: "lazy" | "eager";
};

export type ParseResult<T> =
    | { success: true; data: T }
    | { success: false; error: any };

// Base class for all validators
export abstract class RodType {
    protected def: Record<string, any>;

    // Inject the bridge strategy
    constructor(type: string, protected bridge: Bridge) {
        this.def = { type };
    }

    // Returns the raw JSON spec required by Rust
    toSpec(): object {
        return this.def;
    }

    /**
     * Parse data against the schema. Throws error if validation fails.
     * @param data The input data (unknown)
     * @param opts Configuration for this run
     */
    parse(data: unknown, opts?: ParseOptions): any {
        const schema = this.bridge.getSchema(this.toSpec());
        const mode = opts?.mode ?? "lazy";

        let result;
        if (mode === "lazy") {
            result = schema.validate_lazy(data);
        } else {
            result = schema.validate_eager(data);
        }

        if (!result.success) {
            // Throw error to match Zod behavior
            throw new Error(JSON.stringify(result.error, null, 2));
        }

        return result.data;
    }

    /**
     * Parse data safely without throwing.
     */
    safeParse(data: unknown, opts?: ParseOptions): ParseResult<any> {
        try {
            const res = this.parse(data, opts);
            return { success: true, data: res };
        } catch (e: any) {
            // RodError is structured JSON, so we parse the error message if it was stringified
            let error;
            try {
                error = JSON.parse(e.message);
            } catch {
                error = e.message;
            }
            return { success: false, error };
        }
    }
}

// --- Primitives ---

export class RodString extends RodType {
    constructor(bridge: Bridge) { super("string", bridge); }

    min(val: number) { this.def.min = val; return this; }
    max(val: number) { this.def.max = val; return this; }
    length(val: number) { this.def.length = val; return this; }
    email() { this.def.email = true; return this; }
    url() { this.def.url = true; return this; }
    uuid() { this.def.uuid = true; return this; }
    cuid() { this.def.cuid = true; return this; }
    datetime() { this.def.datetime = true; return this; }
    ip() { this.def.ip = true; return this; }
    regex(pattern: string) { this.def.regex = pattern; return this; }
    startsWith(val: string) { this.def.starts_with = val; return this; }
    endsWith(val: string) { this.def.ends_with = val; return this; }
    includes(val: string) { this.def.includes = val; return this; }
    trim() { this.def.trim = true; return this; }
}

export class RodNumber extends RodType {
    constructor(bridge: Bridge) { super("number", bridge); }

    min(val: number) { this.def.min = val; return this; }
    max(val: number) { this.def.max = val; return this; }
    int() { this.def.int = true; return this; }
}

export class RodBoolean extends RodType {
    constructor(bridge: Bridge) { super("boolean", bridge); }
}

export class RodDate extends RodType {
    constructor(bridge: Bridge) { super("date", bridge); }

    min(timestamp: number) { this.def.min = timestamp; return this; }
    max(timestamp: number) { this.def.max = timestamp; return this; }
}

export class RodLiteral extends RodType {
    constructor(value: any, bridge: Bridge) {
        super("literal", bridge);
        this.def.value = value;
    }
}

export class RodEnum extends RodType {
    constructor(values: string[], bridge: Bridge) {
        super("enum", bridge);
        this.def.values = values;
    }
}

// --- Collections ---

export class RodArray extends RodType {
    constructor(itemSchema: RodType, bridge: Bridge) {
        super("array", bridge);
        this.def.items = itemSchema.toSpec();
    }

    min(val: number) { this.def.min = val; return this; }
    max(val: number) { this.def.max = val; return this; }
}

export class RodObject extends RodType {
    constructor(shape: Record<string, RodType>, bridge: Bridge) {
        super("object", bridge);
        const props: Record<string, any> = {};
        for (const k in shape) {
            props[k] = shape[k].toSpec();
        }
        this.def.properties = props;
    }

    strict() { this.def.strict = true; return this; }
    passthrough() {
        // Note: passthrough support might need explicit logic in your Parser logic in Rust 
        // to map 'strict: false' or a specific flag. Assuming default (strip) vs strict.
        // If Rust parser supports 'passthrough' flag, add it here.
        // For now, based on previous code, we only had strict option exposed in parser.
        // If needed: this.def.passthrough = true; 
        return this;
    }
}

export class RodTuple extends RodType {
    constructor(items: RodType[], bridge: Bridge) {
        super("tuple", bridge);
        this.def.items = items.map(i => i.toSpec());
    }
}

export class RodRecord extends RodType {
    constructor(keySchema: RodType, valueSchema: RodType, bridge: Bridge) {
        super("record", bridge);
        this.def.key = keySchema.toSpec();
        this.def.value = valueSchema.toSpec();
    }
}

export class RodMap extends RodType {
    constructor(keySchema: RodType, valueSchema: RodType, bridge: Bridge) {
        super("map", bridge);
        this.def.key = keySchema.toSpec();
        this.def.value = valueSchema.toSpec();
    }
}

export class RodSet extends RodType {
    constructor(valueSchema: RodType, bridge: Bridge) {
        super("set", bridge);
        this.def.value = valueSchema.toSpec();
    }

    min(val: number) { this.def.min = val; return this; }
}

// --- Logic ---

export class RodUnion extends RodType {
    constructor(options: RodType[], bridge: Bridge) {
        super("union", bridge);
        this.def.options = options.map(o => o.toSpec());
    }
}

export class RodDiscriminatedUnion extends RodType {
    constructor(discriminator: string, options: RodType[], bridge: Bridge) {
        super("discriminatedUnion", bridge);
        this.def.discriminator = discriminator;
        this.def.options = options.map(o => o.toSpec());
    }
}

File: languages/python/rod/builders.py
import json
from typing import Any, Dict, List, Tuple, Union, Literal

# --- Bridge to the Rust native extension ---

# This will be lazily populated to avoid import errors if the native module isn't built yet.
_native_schema = None
_schema_cache: Dict[str, Any] = {}

def _get_compiled_schema(spec: Dict[str, Any]):
    """
    Lazily imports the native Rust module and compiles/caches the schema.
    """
    global _native_schema
    if _native_schema is None:
        try:
            # The native module is named 'rod_py' as defined in Cargo.toml
            from rod.rod_pyo3 import RodSchema
            _native_schema = RodSchema
        except ImportError:
            raise ImportError(
                "The 'rod' native extension is not installed. "
                "Please run 'pip install -e .' or 'maturin develop' in your project root."
            )
    
    # Use a stable JSON string as the cache key
    cache_key = json.dumps(spec, sort_keys=True)
    if cache_key in _schema_cache:
        return _schema_cache[cache_key]
    
    # Compile the schema by calling the Rust constructor
    compiled = _native_schema(spec)
    _schema_cache[cache_key] = compiled
    return compiled

# --- Base Class ---

class RodType:
    """Base class for all Rod schema types."""
    
    def __init__(self, type_name: str):
        self._def: Dict[str, Any] = {"type": type_name}

    def to_spec(self) -> Dict[str, Any]:
        """Returns the raw schema definition dictionary."""
        return self._def

    def parse(self, data: Any) -> Any:
        """
        Validates data against the schema.
        Raises ValueError if validation fails.
        """
        schema = _get_compiled_schema(self.to_spec())
        try:
            return schema.validate(data)
        except ValueError as e:
            # The error from Rust is a JSON string, so we parse it
            # to raise a more structured Python error if needed.
            try:
                error_detail = json.loads(str(e))
                # You could raise a custom RodError exception here
                raise ValueError(error_detail)
            except json.JSONDecodeError:
                raise e # Raise the original error if it wasn't JSON

    def safe_parse(self, data: Any) -> Dict[str, Any]:
        """Validates data safely without throwing an exception."""
        try:
            return {"success": True, "data": self.parse(data)}
        except ValueError as e:
            error_detail = json.loads(str(e)) if isinstance(e.args[0], str) else e.args[0]
            return {"success": False, "error": error_detail}

# --- Primitive Builders ---

class RodString(RodType):
    def __init__(self):
        super().__init__("string")

    def min(self, val: int): self._def["min"] = val; return self
    def max(self, val: int): self._def["max"] = val; return self
    def length(self, val: int): self._def["length"] = val; return self
    def email(self): self._def["email"] = True; return self
    def url(self): self._def["url"] = True; return self
    def uuid(self): self._def["uuid"] = True; return self
    def datetime(self): self._def["datetime"] = True; return self
    def ip(self): self._def["ip"] = True; return self
    def regex(self, pattern: str): self._def["regex"] = pattern; return self
    def trim(self): self._def["trim"] = True; return self

class RodNumber(RodType):
    def __init__(self):
        super().__init__("number")

    def min(self, val: float): self._def["min"] = val; return self
    def max(self, val: float): self._def["max"] = val; return self
    def int(self): self._def["int"] = True; return self

class RodBoolean(RodType):
    def __init__(self):
        super().__init__("boolean")

class RodLiteral(RodType):
    def __init__(self, value: Any):
        super().__init__("literal")
        self._def["value"] = value

class RodEnum(RodType):
    def __init__(self, values: List[str]):
        super().__init__("enum")
        self._def["values"] = values

# --- Collection Builders ---

class RodObject(RodType):
    def __init__(self, shape: Dict[str, RodType]):
        super().__init__("object")
        self._def["properties"] = {k: v.to_spec() for k, v in shape.items()}
    
    def strict(self):
        self._def["strict"] = True
        return self

class RodArray(RodType):
    def __init__(self, item_schema: RodType):
        super().__init__("array")
        self._def["items"] = item_schema.to_spec()

    def min(self, val: int): self._def["min"] = val; return self
    def max(self, val: int): self._def["max"] = val; return self

class RodTuple(RodType):
    def __init__(self, items: Tuple[RodType, ...]):
        super().__init__("tuple")
        self._def["items"] = [item.to_spec() for item in items]

class RodUnion(RodType):
    def __init__(self, options: List[RodType]):
        super().__init__("union")
        self._def["options"] = [opt.to_spec() for opt in options]

class RodDiscriminatedUnion(RodType):
    def __init__(self, discriminator: str, options: List[RodObject]):
        super().__init__("discriminatedUnion")
        self._def["discriminator"] = discriminator
        self._def["options"] = [opt.to_spec() for opt in options]